### PR TITLE
Rule: Errors should not be created without being thrown

### LINKED ISSUE #153: Rule: Exception should not be created without being thrown
Implements [RSPEC-3984](https://jira.sonarsource.com/browse/RSPEC-3984)

### PR DESCRIPTION
**Pull request by inverno**
Fixes #153

### CODE REVIEW

File: sonarts-core/src/rules/noUnthrownErrorRule.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as tslint from "tslint";
+import * as ts from "typescript";
+import { SonarRuleMetaData } from "../sonarRule";
+
+export class Rule extends tslint.Rules.AbstractRule {
+  public static metadata: SonarRuleMetaData = {
+    description: "Error should not be created without being thrown",
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noUnthrownErrorRule.ts]: @inverno title should be in plural number

File: sonarts-core/tests/rules/noUnthrownErrorRule/noUnthrownErrorRule.lint.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// OK
+foo(new Error());
+foo(TypeError);
+throw new Error();
+
+// NOK
+  new Error();
```
Comment (vilchik-elena) [inline sonarts-core/tests/rules/noUnthrownErrorRule/noUnthrownErrorRule.lint.ts]: @inverno error will be created even without `new`. I think we should cover this case as well

File: sonarts-core/src/rules/noUnthrownErrorRule.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      this.raiseIssue(node);
+    }
+  }
+
+  private raiseIssue(node: ts.Node) {
+    this.addFailureAtNode(node, `Throw this error or remove this useless statement`);
+  }
+
+  private isWithinExpressionStatement(node: ts.NewExpression): boolean {
+    if (node.parent) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noUnthrownErrorRule.ts]: @inverno `node.parent  && node.parent.kind === ts.SyntaxKind.ExpressionStatement`? IMO more readable

File: sonarts-core/src/rules/noUnthrownErrorRule.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+class Walker extends tslint.RuleWalker {
+
+  protected visitNewExpression(node: ts.NewExpression): void {
+    if (this.isWithinExpressionStatement(node) && node.expression.getText().endsWith("Error")) {
+      this.raiseIssue(node);
+    }
+  }
+
+  private raiseIssue(node: ts.Node) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noUnthrownErrorRule.ts]: @inverno why this extra function?:)

File: sonarts-core/src/rules/noUnthrownErrorRule.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+  public apply(sourceFile: ts.SourceFile): tslint.RuleFailure[] {
+    return this.applyWithWalker(new Walker(sourceFile, this.getOptions()));
+  }
+}
+
+class Walker extends tslint.RuleWalker {
+
+  protected visitNewExpression(node: ts.NewExpression): void {
+    if (this.isWithinExpressionStatement(node) && node.expression.getText().endsWith("Error")) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noUnthrownErrorRule.ts]: @inverno forgot super

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/resources/org/sonar/l10n/typescript/rules/typescript/S3984.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{
+  "title": "Exception should not be created without being thrown ",
```
Comment (vilchik-elena) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/resources/org/sonar/l10n/typescript/rules/typescript/S3984.json]: @inverno old title

**Comment by vilchik-elena** [comment]
@inverno add description as `md` file

File: sonarts-core/src/rules/noUnthrownErrorRule.ts (hunk 3 above)
Comment (inverno) [reply sonarts-core/src/rules/noUnthrownErrorRule.ts]: @vilchik-elena it requires the cryptic !!node.parent to actually work, but, why not. Done

File: sonarts-core/src/rules/noUnthrownErrorRule.ts (hunk 4 above)
Comment (inverno) [reply sonarts-core/src/rules/noUnthrownErrorRule.ts]: @vilchik-elena habit. Removed.

File: sonarts-core/src/rules/noUnthrownErrorRule.ts (hunk 5 above)
Comment (vilchik-elena) [inline sonarts-core/src/rules/noUnthrownErrorRule.ts]: @inverno let's check also `node.expression.getText().endsWith("Exception")`

Comment (inverno) [reply sonarts-core/src/rules/noUnthrownErrorRule.ts]: Thx, done

Comment (inverno) [reply sonarts-core/src/rules/noUnthrownErrorRule.ts]: Done

File: sonarts-core/src/rules/noUnthrownErrorRule.ts (hunk 1 above)
Comment (inverno) [reply sonarts-core/src/rules/noUnthrownErrorRule.ts]: Done

File: sonarts-core/tests/rules/noUnthrownErrorRule/noUnthrownErrorRule.lint.ts (hunk 2 above)
Comment (inverno) [reply sonarts-core/tests/rules/noUnthrownErrorRule/noUnthrownErrorRule.lint.ts]: @vilchik-elena I'll try and see what shows up with this one

**Pull request merged**